### PR TITLE
Model submission for CTU_CRAS_NORLAB_MARV_SENSOR_CONFIG_1 (and other sensor configs)

### PR DESCRIPTION
**Pull request by peci1**
This PR adds MARV robot (Mobile Autonomous Rescue Vehicle) of team CTU-CRAS-Norlab in 4 sensor configs. The sensor configs differ only in the following features:

1. base config
2. 12 breadcrumbs
3. thermal camera
4. 12 breadcrumbs + thermal camera

MARV robot has not yet finished manufacturing by the company that builds it for us. The expected (and realistic) time to finish it is May 2021. We will provide all validation data as soon as we get the robot driving, which was allowed in an email communication with Angela Maio on 18 Feb 2021:

> model validation data can be submitted up to 3 months after submission of the model (in accordance with the "Model Validation Sessions" contract deliverable)

The design of flippers and robot body have been already finished, while the top deck and sensor box might still undergo minor changes. I'll try to keep the model up-to-date with latest iterations with the manufacturer (until the PR is merged).

We have already uploaded the validation data ZIP and have included a pair of CAD drawings proving conformance of this simulation model with the dimensions of the robot being manufactured.

Flipper control is shared with CTU_CRAS_NORLAB_ABSOLEM robot, so the control plugin is not included and the one from Absolem is used. Once #551 is merged, it will automatically add support for relative positional control in addition to absolute positional control and velocity control available now. #541 adds convenience support for controlling the flippers on D-PAD of gamepads. 

The SDF files should not be edited manually. The model configuration process of MARV starts with Xacro files in `urdf/` folder and config files in `config/`. A combination of these files yields an intermediate URDF, which is converted to SDF via `ign sdf -p` and then polished and postprocessed using `scripts/clean_sdf`. This whole conversion process is done in script `scripts/update_robot_sdf_ign`. If a change should affect all sensor configs of MARV, use `scripts/update_robot_sdf_ign_all_configs` to re-generate all sensor config SDFs. 

There is no URDF file as it is generated on-the-fly by calling `scripts/print_robot_urdf` script. This script is utilized by `launch/description.launch`.

The three additional sensor configs reuse most of the code and configuration of sensor config 1 instead of copy-pasting it.

Manual inspection of the model might be easier in Gazebo Classic. The model contains script `scripts/update_robot_sdf_gz` which overwrites model.sdf with a version compatible with Gazebo Classic. The Classic version is not very well fine-tuned (e.g. it explodes when unpaused), but for static analysis and viewing the model properties and inertia, it works quite well.

### CODE REVIEW

**Comment by nkoenig** [comment]
Bounding box size of this model is `[1.27187 0.573132 0.561826]`

**Review by acschang — CHANGES_REQUESTED** [review]
The initial assessment of the CTU-CRAS-NORLAB MARV model is complete however there are some issues to resolve before this model can be merged and utilized in a competitive setting:

- Please resolve the Ignition to ROS transport issue for MARV sensor configurations 3 and 4.
- Please resolve the Ignition to ROS transport issue for IMU data.
- Please update the IMU parameters to align with the new standard in [798](https://github.com/osrf/subt/pull/798).
- Please update the simulated TFMini sensors to have 9 rays as stated in `specifications.md`.  Currently the sensors are publishing a LaserScan message that only includes 3 rays.
- Please verify that the 400 Hz update rate on the odom topic is intended.
- Please provide additional documentation for the field of view on the [Basler Lens C125-0418-5M-P f4mm](https://www.baslerweb.com/en/products/vision-components/lenses/basler-lens-c125-0418-5m-p-f4mm/).
- Please provide validation data as it is necessary to complete the evaluation. If not provided, the model's maximum velocity will be limited to 1m/s and battery life set to 60min. When the model validation data can be provided later, please submit a PR updating the model.

Please address the issues noted above in this pull request. Once the above issues are fixed, we will evaluate sensor configuration 3 and 4, provide feedback if necessary, and then move your submission to the next review stage.

**Comment by peci1** [comment]
The field of view of the Basler cameras is experimentally measured. We took a ruler, shot a picture of it, and by using a bit of trigonometry, we ended up with horizontal FOV of 86.5°. For the trigonometrical computation we used the datasheet value of "Front focal length", which is 9.95 mm. We sketched a triangle from the point on lens axis that is 9.95 mm inside the lens (measured from the vertex of the outern-most lens). One side of the triangle we used for computation is distance of the ruler from the point inside lens, and the other side is the distance viewed on the ruler.

The theoretical value for a 4 mm lens on 1/2.3" chip is 79°, so I assume the difference to our measured value is done by distortion, which is not modeled in the simulator, and neither is it modeled in the theoretical equation.

The computation I used for horizontal FOV is:

    degrees(2*atan(tan(0.0066)/(2*0.004))) = 79.04 

Where `0.004` is effective focal length and `0.0066` is the sensor width (both from datasheet). This theoretical value is valid for focusing into infinity. Closer focus can decrease the FOV by a few degrees.

**Comment by acschang** [comment]
> The field of view of the Basler cameras is experimentally measured. We took a ruler, shot a picture of it, and by using a bit of trigonometry, we ended up with horizontal FOV of 86.5°. For the trigonometrical computation we used the datasheet value of "Front focal length", which is 9.95 mm. We sketched a triangle from the point on lens axis that is 9.95 mm inside the lens (measured from the vertex of the outern-most lens). One side of the triangle we used for computation is distance of the ruler from the point inside lens, and the other side is the distance viewed on the ruler.
> 
> The theoretical value for a 4 mm lens on 1/2.3" chip is 79°, so I assume the difference to our measured value is done by distortion, which is not modeled in the simulator, and neither is it modeled in the theoretical equation.
> 
> The computation I used for horizontal FOV is:
> 
> ```
> degrees(2*atan(tan(0.0066)/(2*0.004))) = 79.04 
> ```
> 
> Where `0.004` is effective focal length and `0.0066` is the sensor width (both from datasheet). This theoretical value is valid for focusing into infinity. Closer focus can decrease the FOV by a few degrees.

Thank you for the explanation on the sensor FoV.  The experimentally measured value cannot be accepted without real-world documentation such as in [this commit of a related submission](https://github.com/osrf/subt/pull/811/commits/ca4424652046f1ef64269b74205f449a590c2918).  The derived FoV from the specification sheet can be accepted without further documentation (79.04).

**Comment by peci1** [comment]
Okay, we'll deliver the documentation of the experiment.

**Comment by peci1** [comment]
I updated the model and resolved most issues. I also downgraded model.sdf to use SDF 1.6 as requested for all our models.

* [x]  Please resolve the Ignition to ROS transport issue for MARV sensor configurations 3 and 4.
* [x]  Please resolve the Ignition to ROS transport issue for IMU data.
* [x]  Please update the IMU parameters to align with the new standard in [798](https://github.com/osrf/subt/pull/798).
* [x]  Please update the simulated TFMini sensors to have 9 rays as stated in `specifications.md`.  Currently the sensors are publishing a LaserScan message that only includes 3 rays.
* [x]  Please verify that the 400 Hz update rate on the odom topic is intended.
  * That was because all 5 diffdrive plugins were publishing it. I only left one (the one with the largest wheels) as the publisher.
* [ ]  Please provide additional documentation for the field of view on the [Basler Lens C125-0418-5M-P f4mm](https://www.baslerweb.com/en/products/vision-components/lenses/basler-lens-c125-0418-5m-p-f4mm/).
  * We'll provide it as soon as we get back to the lab
* [ ]  Please provide validation data as it is necessary to complete the evaluation. If not provided, the model's maximum velocity will be limited to 1m/s and battery life set to 60min. When the model validation data can be provided later, please submit a PR updating the model.
  * These data will be collected during May. So set the default limits until then.

File: submitted_models/ctu_cras_norlab_marv_sensor_config_1/urdf/boson_thermocam.xacro [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                            <height>256</height>
+                            <format>L16</format>
+                        </image>
+                        <clip>
+                            <near>0.2</near>
+                            <far>50</far>
+                        </clip>
+                    </camera>
+                    <plugin filename="ignition-gazebo-thermal-sensor-system" name="ignition::gazebo::systems::ThermalSensor">
+                        <min_temp>233.15</min_temp>
```
Comment (acschang) [inline submitted_models/ctu_cras_norlab_marv_sensor_config_1/urdf/boson_thermocam.xacro:43]: > Sensor configurations 3 and 4 carry an additional thermocamera FLIR Boson with 92 degrees horizontal FOV lens, modeled by `thermal` sensor. Resolution of the camera is 320x256 px at 8.6 Hz, and temperature range is approximately -50 °C up to 180 °C.

Should this be `223.15` to correspond to -50 degrees C?

Comment (peci1) [reply submitted_models/ctu_cras_norlab_marv_sensor_config_1/urdf/boson_thermocam.xacro:43]: Good catch!

**Comment by peci1** [comment]
Here's the proof of measured FOV of the Basler camera with 4 mm lens: https://drive.google.com/drive/folders/1-KWMvjWSYYeLqPA0C3FCdKuHQAmOHN-z?usp=sharing .

I got the description of our computation wrong before - we did not use front focal length (9.95 mm "inside" the lens), but entrance pupil position (10.66 mm "inside" the lens). According to [wikipedia](https://en.wikipedia.org/wiki/Entrance_pupil):

> The geometric location of the entrance pupil is the vertex of the camera's angle of view[1] and consequently its center of perspective, perspective point, view point, projection centre[2] or no-parallax point.[3] This point is important in panoramic photography, because the camera must be rotated around it in order to avoid parallax errors in the final, stitched panorama.

The value of 10.66 mm is from the lens datasheet: https://docs.baslerweb.com/c125-0418-5m-p .

**Comment by peci1** [comment]
I added camera intrinsics, changed Ouster resolution and stddev to 0.01 and fixed the Boson camera lowest temperature.

**Comment by acschang** [comment]
> Here's the proof of measured FOV of the Basler camera with 4 mm lens: https://drive.google.com/drive/folders/1-KWMvjWSYYeLqPA0C3FCdKuHQAmOHN-z?usp=sharing .
> 
> I got the description of our computation wrong before - we did not use front focal length (9.95 mm "inside" the lens), but entrance pupil position (10.66 mm "inside" the lens). According to [wikipedia](https://en.wikipedia.org/wiki/Entrance_pupil):
> 
> > The geometric location of the entrance pupil is the vertex of the camera's angle of view[1] and consequently its center of perspective, perspective point, view point, projection centre[2] or no-parallax point.[3] This point is important in panoramic photography, because the camera must be rotated around it in order to avoid parallax errors in the final, stitched panorama.
> 
> The value of 10.66 mm is from the lens datasheet: https://docs.baslerweb.com/c125-0418-5m-p .

The experimental data you linked LGTM and the derived calculations align with your originally experimentally measured HFoV.  Can you add this post's information to the `specifications.md`?  Your most recent commits resolve the previously noted issues as well.

**Comment by acschang** [comment]
Continued in #886

**Pull request closed without merging**